### PR TITLE
Warn on type named "record"

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/47090

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs [hunk 1]
```diff
@@ -474,6 +474,7 @@ internal static int GetWarningLevel(ErrorCode code)
                 case ErrorCode.WRN_IsPatternAlways:
                 case ErrorCode.WRN_SwitchExpressionNotExhaustiveWithWhen:
                 case ErrorCode.WRN_SwitchExpressionNotExhaustiveForNullWithWhen:
+                case ErrorCode.WRN_RecordNamedDisallowed:
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs:477]: Shouldn't this warning level be 5 and be introduced as part of the C# 9.0 warning waves?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 2]
```diff
@@ -411,13 +409,18 @@ private DeclarationModifiers MakeModifiers(TypeKind typeKind, DiagnosticBag diag
...
                         {
-                            diagnostics.Add(ErrorCode.ERR_MissingPartial, singleDeclaration.NameLocation, self.Name);
+                            diagnostics.Add(ErrorCode.ERR_MissingPartial, singleDeclaration.NameLocation, this.Name);
                             modifierErrors = true;
                         }
                     }
                 }
             }
 
+            if (this.Name == "record")
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:419]: Should this warning only be issued if the selected language version >= IDS_FeatureRecords.RequiredVersion() ?

File: src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs:477]: No, this is tied to language version, not warn version.

---
In reply to: [476371816](https://github.com/dotnet/roslyn/pull/47094#discussion_r476371816) [](ancestors = 476371816)

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 3]
```diff
@@ -6416,6 +6416,12 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_CloneDisallowedInRecord" xml:space="preserve">
     <value>Members named 'Clone' are disallowed in records.</value>
   </data>
+  <data name="WRN_RecordNamedDisallowed" xml:space="preserve">
+    <value>A type should not be named 'record'.</value>
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:6420]: Consider something like "Types should not be named 'record'."

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 2 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:419]: Yes, this needs to be tied to C# 9.

---
In reply to: [476373766](https://github.com/dotnet/roslyn/pull/47094#discussion_r476373766) [](ancestors = 476373766)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 4]
```diff
@@ -1980,22 +1980,88 @@ class C
...
+                // (6,28): error CS1519: Invalid token ';' in class, record, struct, or interface member declaration
                 //     record M(record r) => r;
                 Diagnostic(ErrorCode.ERR_InvalidMemberDecl, ";").WithArguments(";").WithLocation(6, 28)
                 );
+
+            comp = CreateCompilation(src, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (2,7): error CS8860: A type should not be named 'record'.
+                // class record { }
+                Diagnostic(ErrorCode.WRN_RecordNamedDisallowed, "record").WithArguments("record").WithLocation(2, 7)
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:2007]: This does not match what we agreed to LDM, I believe (or at least, not what I understood and put in the notes). [https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-08-24.md#warnings-on-types-named-record](https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-08-24.md#warnings-on-types-named-record). If we need to discuss whether the warning was specific to C# 9 or was done in prior versions, please send a follow up email.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 5]
```diff
@@ -1980,22 +1980,88 @@ class C
...
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (2,11): warning CS8860: A type should not be named 'record'.
+                // interface record { }
+                Diagnostic(ErrorCode.WRN_RecordNamedDisallowed, "record").WithArguments("record").WithLocation(2, 11)
+                );
+        }
+
+        [Fact, WorkItem(47090, "https://github.com/dotnet/roslyn/issues/47090")]
+        public void TypeNamedRecord_Record()
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:2040]: Please add tests for the other type declarations, such as delegates, type parameters, enums, using aliases, and anything else in the spec that I'm missing off the top of my head.

**Comment by 333fred** [comment]
Done review pass (commit 1)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:419]: consider using something like `this.Name == SyntaxFacts.GetText(SyntaxKind.RecordKeyword)` or `SyntaxFacts.GetContextualKeywordKind(this.Name) == SyntaxKind.RecordKeyword`

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs (hunk 5 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:2040]: Those are good cases. Thanks
`using` aliases are not type declarations but from discussion with Jared, I'm good with blocking them.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs (hunk 4 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:2007]: Chatted about that with Jared, I'll add LangVer gate. Thanks

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 6]
```diff
@@ -2030,12 +2026,152 @@ interface record { }
...
+        {
+            local<record>();
+        }
+    }
+}
+";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (2,9): warning CS8860: Types and aliases should not be named 'record'.
+                // class C<@record> { }
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:2161]: This feels odd to me, if the user has escaped an identifier presumably they are aware of its similarity to a language keyword and do not wish to be bothered about it.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 7]
```diff
@@ -1980,22 +1980,224 @@ class C
...
+        {
+            var src = @"
+using @record = System.Console;
+";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (2,1): hidden CS8019: Unnecessary using directive.
+                // using @record = System.Console;
+                Diagnostic(ErrorCode.HDN_UnusedUsingDirective, "using @record = System.Console;").WithLocation(2, 1),
+                // (2,7): warning CS8860: Types and aliases should not be named 'record'.
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:2091]: This seems incorrect. For example, `using @string = System.Console` is perfectly legal.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs (hunk 5 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:2176]: They're not type declarations, but they're used in places that could be semantically confused for a `record` type declaration.

**Comment by 333fred** [comment]
Done review pass (commit 2)

File: src/Compilers/CSharp/Portable/Binder/Imports.cs [hunk 8]
```diff
@@ -152,7 +152,7 @@ internal string GetDebuggerDisplay()
                         }
 
                         string identifierValueText = usingDirective.Alias.Name.Identifier.ValueText;
-                        SourceMemberContainerTypeSymbol.ReportTypeNamedRecord(identifierValueText, compilation, diagnostics, usingDirective.Alias.Name.Location);
+                        SourceMemberContainerTypeSymbol.ReportTypeNamedRecord(usingDirective.Alias.Name.Identifier.Text, compilation, diagnostics, usingDirective.Alias.Name.Location);
 
                         if (usingAliases != null && usingAliases.ContainsKey(identifierValueText))
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/Imports.cs:157]: Consider moving the `identifierValueText` declaration closer to its usage to make things more clear.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 9]
```diff
@@ -411,16 +409,32 @@ private DeclarationModifiers MakeModifiers(TypeKind typeKind, DiagnosticBag diag
...
+                            diagnostics.Add(ErrorCode.ERR_MissingPartial, singleDeclaration.NameLocation, this.Name);
                             modifierErrors = true;
                         }
                     }
                 }
             }
 
+            if (this.Name == SyntaxFacts.GetText(SyntaxKind.RecordKeyword))
+            {
+                var location = declaration.Declarations[0].NameLocation;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:421]: Would this not issue an error if you had a partial declaration where the "first" declaration was escaped, but the second one wasn't? Consider testing something like:
```cs
partial class @record {}
partial class record {}
```

Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:421]: Actually, I'm not even sure that this code is necessary? It seems that the only direct implementations of this type are `SourceNamedTypeSymbol` (already modified in this PR), `ImplicitNamedTypeSymbol` (which does not have a user-controlled name), and `SimpleProgramNamedTypeSymbol` (also doesn't have a user-controlled name). Do we have a test case that fails if this is removed?

---
In reply to: [477450215](https://github.com/dotnet/roslyn/pull/47094#discussion_r477450215) [](ancestors = 477450215)

**Comment by 333fred** [comment]
Done review pass (commit 4)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 9 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Many tests fail without this logic. This is responsible for checking the name of the type. The logic added in `SourceNamedTypeSymbol` is responsible for checking type parameters.

For instance this test fails:
```
        public void TypeNamedRecord()
        {
            var src = @"
class record { }

class C
{
    record M(record r) => r;
}";
            var comp = CreateCompilation(src);
...
```

---
In reply to: [477453969](https://github.com/dotnet/roslyn/pull/47094#discussion_r477453969) [](ancestors = 477453969,477450215)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 10]
```diff
@@ -1980,22 +1980,335 @@ class C
...
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (2,15): warning CS8860: Types and aliases should not be named 'record'.
+                // partial class record { }
+                Diagnostic(ErrorCode.WRN_RecordNamedDisallowed, "record").WithArguments("record").WithLocation(2, 15)
+                );
+        }
+
+        [Fact, WorkItem(47090, "https://github.com/dotnet/roslyn/issues/47090")]
+        public void TypeNamedRecord_BothEscapedPartial()
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:2300]: Nit: this is not both escaped :)

**Comment by jcouv** [comment]
@jaredpar @cston for another review. Thanks

File: src/Compilers/CSharp/Portable/Binder/Imports.cs [hunk 11]
```diff
@@ -151,6 +151,8 @@ internal string GetDebuggerDisplay()
                             diagnostics.Add(ErrorCode.ERR_NoAliasHere, usingDirective.Alias.Name.Location);
                         }
 
+                        SourceMemberContainerTypeSymbol.ReportTypeNamedRecord(usingDirective.Alias.Name.Identifier.Text, compilation, diagnostics, usingDirective.Alias.Name.Location);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Imports.cs:154]: >usingDirective.Alias.Name [](start = 78, length = 25)

Consider extracting a local.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 12]
```diff
@@ -411,16 +409,35 @@ private DeclarationModifiers MakeModifiers(TypeKind typeKind, DiagnosticBag diag
...
                     }
                 }
             }
 
+            if (this.Name == SyntaxFacts.GetText(SyntaxKind.RecordKeyword))
+            {
+                foreach (var decl in declaration.Declarations)
+                {
+                    var location = decl.NameLocation;
+                    var nameText = location.SourceTree.GetRoot().FindToken(location.SourceSpan.Start).ToString();
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:424]: >location.SourceTree.GetRoot().FindToken(location.SourceSpan.Start).ToString(); [](start = 35, length = 78)

This seems too heavyweight. Could we use direct access instead? Perhaps `((decl as SingleTypeDeclaration)?.SyntaxReference.GetSyntax() as BaseTypeDeclarationSyntax)?.Identifier.ValueText`.

Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:424]: Or change the loop to loop over `SyntaxReferences`:
```
foreach (var syntaxRef in SyntaxReferences)
{
    var location = syntaxRef.GetLocation();
    var nameText = (syntaxRef as BaseTypeDeclarationSyntax)?.Identifier.ValueText;
    ...
}
```

---
In reply to: [478653015](https://github.com/dotnet/roslyn/pull/47094#discussion_r478653015) [](ancestors = 478653015)

**Pull request merged**